File: Chagok/frontend/src/types/message.ts
/**
 * Message types for real-time communication
 * 003-role-based-ui Feature - US6
 */

// ============== API Types ==============

export interface MessageSender {
  id: string;
  name: string;
  role: string;
}

export interface Message {
  id: string;
  case_id: string;
  sender: MessageSender;
  recipient_id: string;
  content: string;
  attachments?: string[];
  read_at?: string;
  created_at: string;
  is_mine: boolean;
}

export interface ConversationSummary {
  case_id: string;
  case_title: string;
  other_user: {
    id: string;
    name: string;
    role: string;
  };
  last_message: string;
  last_message_at: string;
  unread_count: number;
}

export interface MessageListResponse {
  messages: Message[];
  total: number;
  has_more: boolean;
}

export interface ConversationListResponse {
  conversations: ConversationSummary[];
  total_unread: number;
}

export interface UnreadCountResponse {
  total: number;
  by_case: Record<string, number>;
}

// ============== Request Types ==============

export interface SendMessageRequest {
  case_id: string;
  recipient_id: string;
  content: string;
  attachments?: string[];
}

export interface MarkReadRequest {
  message_ids: string[];
}

// ============== WebSocket Types ==============

export type WebSocketMessageType =
  | 'message'
  | 'read'
  | 'typing'
  | 'ping'
  | 'new_message'
  | 'message_sent'
  | 'read_confirmed'
  | 'read_receipt'
  | 'offline_messages'
  | 'presence'
  | 'pong'
  | 'error';

export interface WebSocketMessage<T = unknown> {
  type: WebSocketMessageType;
  payload: T;
}

export interface TypingPayload {
  user_id: string;
  case_id: string;
  is_typing: boolean;
}

export interface PresencePayload {
  user_id: string;
  status: 'online' | 'away' | 'offline';
}

export interface OfflineMessagesPayload {
  messages: Message[];
}

// ============== UI State Types ==============

export interface MessageState {
  messages: Message[];
  isLoading: boolean;
  error: string | null;
  hasMore: boolean;
}

export interface ConversationState {
  conversations: ConversationSummary[];
  totalUnread: number;
  isLoading: boolean;
  error: string | null;
}

export interface TypingIndicator {
  userId: string;
  caseId: string;
  timestamp: number;
}

// ============== Direct Message Types (US2 - Lawyer Portal) ==============

/**
 * Direct message between users (not case-specific)
 * Used in /lawyer/messages page
 */
export interface DirectMessage {
  id: string;
  senderId: string;
  senderName: string;
  recipientId: string;
  recipientName: string;
  subject: string;
  content: string;
  isRead: boolean;
  createdAt: string; // ISO 8601
  readAt?: string;
}

export interface DirectMessageSummary {
  id: string;
  senderId: string;
  senderName: string;
  recipientId: string;
  recipientName: string;
  subject: string;
  preview: string; // First 100 chars of content
  isRead: boolean;
  createdAt: string;
}

export interface DirectMessageListResponse {
  messages: DirectMessageSummary[];
  total: number;
  page: number;
  limit: number;
}

export interface DirectMessageQueryParams {
  folder: 'inbox' | 'sent';
  page?: number;
  limit?: number;
}

export interface DirectMessageCreate {
  recipientId: string;
  subject: string;
  content: string;
}

export interface DirectMessageState {
  messages: DirectMessageSummary[];
  total: number;
  currentMessage: DirectMessage | null;
  isLoading: boolean;
  error: string | null;
}

File: Chagok/backend/app/schemas/message.py
"""
Message schemas for real-time communication
003-role-based-ui Feature - US6
"""

from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime


# ============== Request Schemas ==============

class MessageCreate(BaseModel):
    """Schema for creating a new message"""
    case_id: str = Field(..., description="Case ID the message belongs to")
    recipient_id: str = Field(..., description="Recipient user ID")
    content: str = Field(..., min_length=1, max_length=5000, description="Message content")
    attachments: Optional[List[str]] = Field(None, description="List of attachment URLs")


class MessageMarkRead(BaseModel):
    """Schema for marking messages as read"""
    message_ids: List[str] = Field(..., description="List of message IDs to mark as read")


# ============== Response Schemas ==============

class MessageSender(BaseModel):
    """Sender information in message"""
    id: str
    name: str
    role: str

    class Config:
        from_attributes = True


class MessageResponse(BaseModel):
    """Single message response"""
    id: str
    case_id: str
    sender: MessageSender
    recipient_id: str
    content: str
    attachments: Optional[List[str]] = None
    read_at: Optional[datetime] = None
    created_at: datetime
    is_mine: bool = False  # Whether the current user is the sender

    class Config:
        from_attributes = True


class OtherUserInfo(BaseModel):
    """Other user info in conversation"""
    id: str
    name: str
    role: str


class ConversationSummary(BaseModel):
    """Summary of a conversation thread"""
    case_id: str
    case_title: str
    other_user: OtherUserInfo
    last_message: str
    last_message_at: datetime
    unread_count: int


class MessageListResponse(BaseModel):
    """List of messages response"""
    messages: List[MessageResponse]
    total: int
    has_more: bool


class ConversationListResponse(BaseModel):
    """List of conversations response"""
    conversations: List[ConversationSummary]
    total_unread: int


class UnreadCountResponse(BaseModel):
    """Unread message count response"""
    total: int
    by_case: dict[str, int]  # case_id -> count


# ============== WebSocket Schemas ==============

class WebSocketMessage(BaseModel):
    """WebSocket message format"""
    type: str = Field(..., description="Message type: message, read, typing, presence")
    payload: dict = Field(..., description="Message payload")


class NewMessagePayload(BaseModel):
    """Payload for new message via WebSocket"""
    case_id: str
    recipient_id: str
    content: str
    attachments: Optional[List[str]] = None


class TypingPayload(BaseModel):
    """Payload for typing indicator"""
    case_id: str
    recipient_id: str
    is_typing: bool


class PresencePayload(BaseModel):
    """Payload for presence update"""
    status: str  # online, away, offline
